### ISSUE TITLE
Prediction Problem

### ISSUE THREAD

**Issue by sulaimanvesal** [rajshah4/image_keras#1]
Dear Rajsh;

Thanks for your nice material. I could figure out how to run the code by Fchollet and everything is fine. Without using fine-tuning and on the entire Dataset of Cat-&-Dog in 50 epochs, I could get accuracy of 98 %. But the problem is with predication on 12500 test images. At first I just predict test-images by changing the shape and size and after loading the model I did prediction which gives dimension error.

    img2 = load_img('Dataset/Test/Test/1.jpg', False, (150 img_150))
    x2 = img_to_array(img2)
    pre = model.predict(x2.reshape((1, 3, 150, 150)), batch_size=32, verbose=1)

    Exception: Error when checking : expected flatten_input_1 to have shape (None, 512, 4, 4) 
    but got   array with shape (1, 3, 150, 150)

Then, I figured out that I should make a generator for test-images and feed the model based on that and then do prediction. I could get some results but predication accuracy is really worse.

    train_data = np.load(open('bottleneck_features_train.npy'))
    train_labels = np.array([0] * (nb_train_samples / 2) + [1] * (nb_train_samples / 2))

    validation_data = np.load(open('bottleneck_features_validation.npy'))
    validation_labels = np.array([0] * (nb_validation_samples / 2) + [1] * (nb_validation_samples / 2))

    test_data = np.load(open('bottleneck_features_test.npy'))

    model = Sequential()
    model.add(Flatten(input_shape=train_data.shape[1:]))
    model.add(Dense(256, activation='relu'))
    model.add(Dropout(0.5))
    model.add(Dense(1, activation='sigmoid'))

    model.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['accuracy'])

    model.fit(train_data, train_labels,
              nb_epoch=nb_epoch, batch_size=32,
              validation_data=(validation_data, validation_labels))

    model.save_weights(top_model_weights_path)

    loss_and_metrics = model.evaluate(validation_data, validation_labels, batch_size=32)
    print (loss_and_metrics)

    predictions = model.predict(test_data, batch_size=32, verbose=1)
    print (predictions)
    proba = model.predict_classes(test_data, batch_size=32, verbose=1)
    print (proba)

    train_id = range(1, 12501)
    print(len(train_id))
    print(predictions[:, 0].shape)
    predictions_df = pd.DataFrame({'id': train_id, 'label': predictions[:, 0]})
    predictions_df.to_csv("Dataset/Test/kagglePridications.csv", index=False)

Would you please help me that what is the problem with this.

**Comment by rajshah4** [comment]
Let me make sure I understand.  You loaded all the cat/dog pics from Kaggle (~12,500) and then used a pretrained network and built a small network on top of it.  If you predict a single image, are the predictions making sense?  What is the prediction accuracy you are seeing (is it like 50% or are you expecting 98% and you are getting 94%)?

**Comment by sulaimanvesal** [comment]
Hi;
You are right. I built a small network on top of pretrained network. As I mentioned there are two cases:
1- If I predict a single image which is not possible because the model only understand the dimension (None, 512, 4, 4) and I don't know how to reshape my test image which has size (1, 3, 150, 150).

2- To resolve this issue, I found that I can make a data generator for my test set (12500 images) which give me same dimension as model. First, I evaluated my model with training data and validation data( Training Data = 25000 images, Validation Data= 5000 Images (20%)). The validation accuracy is really good and could get 98 % after 50 Epochs. In the last, I loaded my test-data which generated by data-generator. The code is working well, but the result of predication is not good, it's like 60 % and many of test-images wrongly classified. 

I hope I could explain and clear the issue.

**Comment by rajshah4** [comment]
If you look at the notebook_extras in the repo, I was able to do prediction using just one image.  Otherwise, just feed your model one class (just cat pictures) and see what is going on.  My guess this is something going on with how you are doing the prediction.  (Can you look at your test data and make sure the labels/images are matching up right that its feeding into the network?)  What happens when you do your predict on the validation data set?  What kind of scores do you see?  That might help your troubleshoot this as well.

**Comment by sulaimanvesal** [comment]
Thank you for your hints. I can predict using one image but not when I use pre-trained model. As observed you also just predict the single image on first model which is not pre-trained. I am going to paste my entire model, if you could help me because it took like two weeks that I am struggling. 

    import os
    import h5py
    import numpy as np
    import pandas as pd
    from PIL import Image
    from keras.preprocessing.image import ImageDataGenerator,load_img,img_to_array
    from keras.models import Sequential
    from keras.layers import Convolution2D, MaxPooling2D, ZeroPadding2D
    from keras.layers import Activation, Dropout, Flatten, Dense
    from keras import backend as K
    K.set_image_dim_ordering('th')

    # path to the model weights file.
    weights_path = 'VGG16Weights/vgg16_weights.h5'
    top_model_weights_path = 'VGG16Weights/bottleneck_fc_model.h5'
    # dimensions of our images.
    img_width, img_height = 150, 150

    train_data_dir = 'Dataset/Train'
    validation_data_dir = 'Dataset/Validation'
    test_data_dir = 'Dataset/Test'
    nb_train_samples = 25000
    nb_validation_samples = 5000
    nb_epoch = 1

    def save_bottlebeck_features():
     datagen = ImageDataGenerator(rescale=1./255)
     # build the VGG16 network
     model = Sequential()
     model.add(ZeroPadding2D((1, 1), input_shape=(3, img_width, img_height)))

    model.add(Convolution2D(64, 3, 3, activation='relu', name='conv1_1'))
    model.add(ZeroPadding2D((1, 1)))
    model.add(Convolution2D(64, 3, 3, activation='relu', name='conv1_2'))
    model.add(MaxPooling2D((2, 2), strides=(2, 2)))

    model.add(ZeroPadding2D((1, 1)))
    model.add(Convolution2D(128, 3, 3, activation='relu', name='conv2_1'))
    model.add(ZeroPadding2D((1, 1)))
    model.add(Convolution2D(128, 3, 3, activation='relu', name='conv2_2'))
    model.add(MaxPooling2D((2, 2), strides=(2, 2)))

    model.add(ZeroPadding2D((1, 1)))
    model.add(Convolution2D(256, 3, 3, activation='relu', name='conv3_1'))
    model.add(ZeroPadding2D((1, 1)))
    model.add(Convolution2D(256, 3, 3, activation='relu', name='conv3_2'))
    model.add(ZeroPadding2D((1, 1)))
    model.add(Convolution2D(256, 3, 3, activation='relu', name='conv3_3'))
    model.add(MaxPooling2D((2, 2), strides=(2, 2)))

    model.add(ZeroPadding2D((1, 1)))
    model.add(Convolution2D(512, 3, 3, activation='relu', name='conv4_1'))
    model.add(ZeroPadding2D((1, 1)))
    model.add(Convolution2D(512, 3, 3, activation='relu', name='conv4_2'))
    model.add(ZeroPadding2D((1, 1)))
    model.add(Convolution2D(512, 3, 3, activation='relu', name='conv4_3'))
    model.add(MaxPooling2D((2, 2), strides=(2, 2)))

    model.add(ZeroPadding2D((1, 1)))
    model.add(Convolution2D(512, 3, 3, activation='relu', name='conv5_1'))
    model.add(ZeroPadding2D((1, 1)))
    model.add(Convolution2D(512, 3, 3, activation='relu', name='conv5_2'))
    model.add(ZeroPadding2D((1, 1)))
    model.add(Convolution2D(512, 3, 3, activation='relu', name='conv5_3'))
    model.add(MaxPooling2D((2, 2), strides=(2, 2)))

    # load the weights of the VGG16 networks
    # (trained on ImageNet, won the ILSVRC competition in 2014)
    # note: when there is a complete match between your model definition
    # and your weight savefile, you can simply call model.load_weights(filename)
    assert os.path.exists(weights_path), 'Model weights not found (see "weights_path" variable in script).'
    f = h5py.File(weights_path)
    for k in range(f.attrs['nb_layers']):
        if k >= len(model.layers):
            # we don't look at the last (fully-connected) layers in the savefile
            break
        g = f['layer_{}'.format(k)]
        weights = [g['param_{}'.format(p)] for p in range(g.attrs['nb_params'])]
        model.layers[k].set_weights(weights)
    f.close()
    print('Model loaded.')

    generator = datagen.flow_from_directory(
            train_data_dir,
            target_size=(img_width, img_height),
            batch_size=32,
            class_mode=None,
            shuffle=False)
    bottleneck_features_train = model.predict_generator(generator, nb_train_samples)
    np.save(open('bottleneck_features_train.npy', 'w'), bottleneck_features_train)

    generator = datagen.flow_from_directory(
            validation_data_dir,
            target_size=(img_width, img_height),
            batch_size=32,
            class_mode=None,
            shuffle=False)
    bottleneck_features_validation = model.predict_generator(generator, nb_validation_samples)
    np.save(open('bottleneck_features_validation.npy', 'w'), bottleneck_features_validation)

    generator = datagen.flow_from_directory(
            test_data_dir,
            target_size=(img_width, img_height),
            batch_size=32,
            shuffle=False)
    test_data_features = model.predict_generator(generator, 12500)
    np.save(open('bottleneck_features_test.npy', 'w'), test_data_features)

    def train_top_model():
    train_data = np.load(open('bottleneck_features_train.npy'))
    train_labels = np.array([0] * (nb_train_samples / 2) + [1] * (nb_train_samples / 2))

    validation_data = np.load(open('bottleneck_features_validation.npy'))
    validation_labels = np.array([0] * (nb_validation_samples / 2) + [1] * (nb_validation_samples / 2))

    test_data = np.load(open('bottleneck_features_test.npy'))

    model = Sequential()
    model.add(Flatten(input_shape=train_data.shape[1:]))
    model.add(Dense(256, activation='relu'))
    model.add(Dropout(0.5))
    model.add(Dense(1, activation='sigmoid'))

    model.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['accuracy'])

    model.fit(train_data, train_labels,
              nb_epoch=nb_epoch, batch_size=32,
              validation_data=(validation_data, validation_labels))

    model.save_weights(top_model_weights_path)

    loss_and_metrics = model.evaluate(validation_data, validation_labels, batch_size=32)
    print (loss_and_metrics)

    predictions = model.predict(test_data, batch_size=32, verbose=1)
    print (predictions)
    proba = model.predict_classes(test_data, batch_size=32, verbose=1)
    print (proba)

    train_id = range(1, 12501)
    print(len(train_id))
    print(predictions[:, 0].shape)
    predictions_df = pd.DataFrame({'id': train_id, 'label': predictions[:, 0]})
    predictions_df.to_csv("Dataset/Test/kagglePridications.csv", index=False)

    #===============================================================================
   save_bottlebeck_features()
   train_top_model()

**Comment by rajshah4** [comment]
Ok, I will try to take a look.  The original post has info on taking a picture and putting it through the image generator, I wonder if that would work for you: 
```
img = load_img('data/train/cats/cat.0.jpg')  # this is a PIL image
x = img_to_array(img)  # this is a Numpy array with shape (3, 150, 150)
x = x.reshape((1,) + x.shape)  # this is a Numpy array with shape (1, 3, 150, 150)

# the .flow() command below generates batches of randomly transformed images
# and saves the results to the `preview/` directory
i = 0
for batch in datagen.flow(x, batch_size=1,
                          save_to_dir='preview', save_prefix='cat', save_format='jpeg'):
    i += 1
    if i > 20:
        break  # otherwise the generator would loop indefinitely
```

**Comment by rajshah4** [comment]
Can you try running a predict on the train set  . . . If that doesn't give high scores than the problem is how the prediction is being done.

**Comment by sulaimanvesal** [comment]
Yes I can run, and that gives really good score. I don't know what's wrong with test data. I thought validation data and train data didn't shuffle very well, but there is no problem with this fact. Here is the predication probability for training images, which is quite good, I am going to paste based on image ID.

12495,0.29315957427
12496,0.0002777276095
12497,0.0614244751632
12498,0.105285122991
12499,0.149591609836
12500,0.0710603222251
12501,0.973863184452
12502,0.97020560503
12503,1.0
12504,0.70070117712
12505,0.725467503071
12506,0.999975919724
12507,0.890727818012
12508,0.887957215309
12509,0.999996423721
12510,0.999843001366
12511,0.982473492622

**Comment by rajshah4** [comment]
Ahh good, that means the model is working just fine. You just need to figure out what is going on with your test_data.  Try just taking a very small selection of images that are all one class and run them through the model as test_data.

**Comment by sulaimanvesal** [comment]
What do you think about this:

    Found 2048 images belonging to 2 classes.
    Found 832 images belonging to 2 classes.
    Found 12500 images belonging to 1 classes.

Don't you think the generator for the test class that feed the images (12500) is going to one-class whether it's not belong to any other class. I am sure this the main problem.

Do you know any alternarive way generate a generator which gives final output after pre-training in shape of (Non, 512,4,4).

**Comment by rajshah4** [comment]
Take a look at: https://keras.io/preprocessing/image/ under flow_from_directory(directory) and the option for class_mode
Also see the example here at the end: https://github.com/fastai/courses/blob/master/deeplearning1/nbs/lesson1.ipynb
```
class_mode: one of "categorical", "binary", "sparse" or None. Default: "categorical". 
Determines the type of label arrays that are returned: "categorical" will be 2D one-hot encoded labels, "binary" will be 1D binary labels, "sparse" will be 1D integer labels. 
If None, no labels are returned (the generator will only yield batches of image data, which is useful to use  model.predict_generator(), model.evaluate_generator(), etc.).
```

**Comment by rajshah4** [comment]
Closing because I think the issue was due to a bad test dataset
